napari/_qt/qt_main_window.py

```diff
@@ -635,6 +635,68 @@ def remove_dock_widget(self, widget: QWidget):
         # rather than directly using _add_viewer_dock_widget
         _dw.deleteLater()
 
+    def add_function_widget(
+        self,
+        function,
+        *,
+        magic_kwargs=None,
+        name: str = '',
+        area: str = 'bottom',
+        allowed_areas=None,
+        shortcut=None,
+    ):
+        """Turn a function into a dock widget via magicgui.
+
+        Parameters
+        ----------
+        function : callable
+            Function that you want to add.
+        magic_kwargs : dict, optional
+            Keyword arguments to the magicgui function that
+            can be used to specify widget.
+        name : str, optional
+            Name of dock widget to appear in window menu.
+        area : str
+            Side of the main window to which the new dock widget will be added.
+            Must be in {'left', 'right', 'top', 'bottom'}
+        allowed_areas : list[str], optional
+            Areas, relative to main window, that the widget is allowed dock.
+            Each item in list must be in {'left', 'right', 'top', 'bottom'}
+            By default, all areas are allowed.
+        shortcut : str, optional
+            Keyboard shortcut to appear in dropdown menu.
+
+        Returns
+        -------
+        dock_widget : QtViewerDockWidget
+            `dock_widget` that can pass viewer events.
+        """
+        from magicgui import magicgui
+
+        if name == '':
+            function_name = function.__name__.replace('_', ' ')
+        else:
+            function_name = name
+
+        if magic_kwargs is None:
+            magic_kwargs = {}
+
+        # Get widget from magicgui
+        widget = magicgui(**magic_kwargs)(function)
+
+        # Keep the dropdown menus in the widget in sync with the layer model
+        self.qt_viewer.viewer.layers.events.connect(
+            lambda x: widget.reset_choices()
+        )
+
+        return self.add_dock_widget(
+            widget,
+            name=function_name,
+            area=area,
+            allowed_areas=allowed_areas,
+            shortcut=shortcut,
+        )
```

Comment:
here's a thought: we could modify `add_dock_widget` to connect the `layers.events` to `widget.reset_choices` when widget is an instance of `magicgui.widgets.Widget`.  That would let us remove that whole "Keep the dropdown menus in the widget in sync" line from _all_ of the examples that use it... and with a little one-liner magic, would reduce this whole function to:

```suggestion
        from magicgui import magicgui

        return self.add_dock_widget(
            magicgui(**magic_kwargs or {})(function),
            name=name or function.__name__.replace('_', ' '),
            area=area,
            allowed_areas=allowed_areas,
            shortcut=shortcut,
        )
```

Reply:
makes sense to me, will do tomorrow